src/Sarif/FileEncoding.cs

```diff
@@ -0,0 +1,101 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Text;
+
+namespace Microsoft.CodeAnalysis.Sarif
+{
+    public class FileEncoding
```

Comment:
>FileEncoding [](http://example.com/codeflow?start=17&length=12)

There's nothing in here yet really about detecting file encodings, a very hard problem. We strictly attempt classification but the use of of three specific encodings is critical to this.

We use UTF8 as a general mechanism for sniffing ASCII and various Unicode. We use Windows code page as this CP allows for certain extended ASCII characters that are replaced when decoded as UTF8. Finally, UTF32 (an unusual encoding) is handled specially so that we can use the trick of classifying data as binary when it contains '\0' chars.

Currently, I've used this to classify around 175k files from the world's largest GitHub repository. It's done a pretty impressive job of sorting files for this data set. I found many examples of files whose extension didn't match reality (mostly test data). 

One minor concern, sometimes developers save source code with the Unicode replacement char in it (due to round-tripping the file, perhaps, in an editor that doesn't support some copied text, typically in a code comment). 

We could attempt to address this moving forward by other heuristics, we could provide an extensions list ('always treat these as text/binary') or we could attempt to compute a % of characters in the sniff that were invalid.

I continue to be concerned about the perf impact of introducing this helper in a general scan framework. However, it has also occurred to me that some scanners could be attempting to scan large amounts of binary data as textual (this certainly happens in one of our current tools). And so this classifier might allow all of that analysis to go away. We'll see. I'll test real world data sets today.